Replace @wordpress/components Button, Radio, RadioGroup with Ariakit Button

samueljseay opened the pull request:
### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/46062 https://github.com/woocommerce/woocommerce/issues/46208

This introduces the Ariakit `Button` component. We use it to replace our use of the `@wordpress/components` button, but also to replace `__experimentalRadioGroup` and `__experimentalRadio`.

I considered using Ariakit `RadioGroup` and `Radio` for these, here's a summary of my reasoning not to:

* They are used for one single feature in checkout, selecting between pickup and shipping, and their styling / functionality is more like 2 giant buttons that select between 2 values managed in state, not actual form data. For aria we can just apply roles of `radiogroup` and `radio`.

* We already have some existing state management of 2 simple values, if we used RadioGroup and Radio we'd need to introduce not only RadioGroup, Radio but also the store and state management for those components and we just don't need them. Furthermore this is the **only** instance where we use the old wordpress components for pseudo radio functionality, so introducing a whole new dependency here is just not worth it.

* Also if we did introduce Radio/RadioGroup styling them was going to be a lot more work, they are radio buttons by default, whereas with using buttons here we get to make very few changes to get button styles  because the wordpress component (which is now deprecated) was also using buttons under the hood.

#### Deprecation of showSpinner

In the past on WooCommerce we've tended to have API's like `showSpinner` for components that will trigger the rendering or not rendering of some child component like a spinner. I've deprecated this because I think it's always better to render something like that in `children` of the component. It's more flexible in general and especially with a primitive component it is better not to render any child content in the component by default because it's not typically needed. It also allows for better performance in future, if we don't bundle components together in this way, if an entrypoint uses button but never with a spinner, for example, it should not need a dependency on `Spinner`.

To achieve the same result you can render <Spinner> as a child when you have a loading state.

e.g.

```
<Button>
   {isLoading && <Spinner />}
</Button>
```

#### Removing support for icon

For the same reasoning as `showSpinner` we don't need to introduce opinionated ways that children are rendered in primitive components. One can always use composition in future, if rendering a button with an icon is needed (ie create a IconButton component). 

Furthermore, we never documented that the API of `@wordpress/components` button was the public API of our component, it just so happens in a few places we relied on the passing down of all props to use the icon prop, but there are other props like `iconPosition` and others that one could technically use in the current implementation. I think it's better to remove usage of any underlying props of wp button instead of trying to support backwards compat for them as we never advertised they were part of the API.

To achieve the same result of rendering an icon as before, you should render it as a child of the button as we've done in this PR.

e.g.

```
<Button>
  <Icon icon={ close } />
</Button>
```

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

* Mini cart close button (top right of open mini-cart)
* All buttons and link buttons in the cart and checkout
* Close button on the notice component
* Local pickup / Shipping "buttons" in the checkout form

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
In blocks migrate `@wordpress/components` Button to Ariakit, replace  `__experimentalRadio/RadioGroup` with Ariakit Button.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

samueljseay committed:
WIP vendoring and consolidating bundles.

samueljseay committed:
Enable externalizing blocks components and checkout from frontend config.

samueljseay committed:
Target all node modules.

samueljseay committed:
Don't vendor split for product-button-interactivity, to avoid introducing issues.

samueljseay committed:
Give each bundle a unique name to ensure no runtime conflicts between bundles.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

samueljseay committed:
Reintroduce fix for product-button-interactivity.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/attempt-vendoring

samueljseay committed:
Fix flakyness in cart/checkout translation e2e tests.

samueljseay committed:
Adjust file naming to match old bundle to avoid breaking translations.

samueljseay committed:
WIP introduce ariakit button.

samueljseay committed:
Introduce Ariakit button component.

samueljseay committed:
Revert redundant changes to place order button.

samueljseay committed:
Revert changes to package json.

opr commented:
@samueljseay FYI I am marking this as blocked, pending the merge of https://github.com/woocommerce/woocommerce/pull/45859

samueljseay committed:
Migrate the radio group to a simple pair of ariakit buttons.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Merge branch 'dev/migrate-radiogroup-to-button' into dev/introduce-ariakit-button

samueljseay committed:
Use onclick to change between shipping and pickup.

samueljseay committed:
Remove another use of important we don't need now.

samueljseay committed:
Add both selected styles to focus state to ensure it doesn't override them.

samueljseay commented on plugins/woocommerce-blocks/assets/js/base/components/button/index.tsx:
```diff
@@ -23,48 +31,77 @@ export interface ButtonProps
+			variant = 'contained',
+			// To maintain backward compat we render a wrapper for button text by default,
+			// but you can opt in to removing it by setting removeTextWrap to true.
+			removeTextWrap = false,
```
I think in general it would be nicer for consumers of Button to do things like adding a text wrapper on children if needed, but for simplicity of migration to the new component I opted to leave things as they were before. I just need to opt out of rendering the wrapper for the buttons used in the radio button scenario.

samueljseay committed:
Use boolean shorthand on button components.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

samueljseay committed:
Ensure style specificity, remove unused text wrapper on mini cart close button.

samueljseay committed:
Render icon as child where relevant, replace label with aria label.

samueljseay commented on plugins/woocommerce-blocks/assets/js/base/components/drawer/index.tsx:
```diff
@@ -56,11 +56,12 @@ const CloseButtonPortal = ( {
+					removeTextWrap
+					aria-label={ __( 'Close', 'woocommerce' ) }
+				>
+					<Icon icon={ close } />
```
I have opted not to support the `icon` arg. But of course if we think we should retain it for backwards compat I'm open to it.

These APIs are definitely bad though, most things you can do by rendering a child instead of passing a prop and its almost always much more flexible, reusable etc.

opr replied:
Do you think we should keep it in, but make it show a deprecation notice along with instructions on the better solution? Would it be too complex to support both approaches at the same time?

samueljseay replied:
Not too complex. Happy to add deprecation instead :)

samueljseay replied:
@opr I just looked a little bit deeper into this, and I just want to maybe challenge the idea to support `icon` here and see what you think.

The reason `icon` works is that it is part of the underlying `Button` API of the wordpress component. So we were just passing it through along with any other prop that WP button supports. We used this internally, but we never documented it in Storybook and I don't see any documentation that implies you can just use any prop supported by WP button.. well that is except for the TS type we use for the props, but that was fairly recent and more a result of converting to TS.

If we continue to support `icon` I think it implies we'd support the rest of the icon api supported on WP button as well, such as `iconPosition` and `iconSize`. We'd also be supporting it only for anyone who externally consumes this component, because as you can see here it was trivial for me to just render an icon in the children, so internally we don't need this.

I think we should just ditch support for passing through any of the old props of WP button because I think you could argue if we don't support all of them with deprecation warnings we shouldn't support any of them.

I would prefer to just support the obviously exposed props such as `showSpinner` (which I have). Wdyt?

I also think we could/should deprecate `showSpinner` because its another situation where composing components is better, but I won't suggest it here because unless someone gets given the task of working on f/e components framework I think that deprecation warning will stick around a long time.

opr replied:
Thanks for the thoughts here, it makes sense to discontinue support for the WP button props, and I think it falls outside of our usual way of deprecating props (given they were never advertised). That said, we should make sure we announce this ahead of the release that includes this change, as part of a developer advisory. In fact, we could even do that now in a discussion. What do you think?

I agree `showSpinner` is a little more complicated to remove because we have advertised it in documentation (Storybook) but you're right, composition makes more sense. 

>  I won't suggest it here because unless someone gets given the task of working on f/e components framework I think that deprecation warning will stick around a long time.

I don't know if the deprecation notice being around for a long time is really a problem. Arguably, the longer it's around for the better. It gives folks more time to move away from it, maybe we _should_ add it as part of this effort?

samueljseay replied:
> I don't know if the deprecation notice being around for a long time is really a problem. Arguably, the longer it's around for the better. It gives folks more time to move away from it, maybe we should add it as part of this effort?

Yep I can get behind that argument. I'll add a deprecation here.

> That said, we should make sure we announce this ahead of the release that includes this change, as part of a developer advisory. In fact, we could even do that now in a discussion. What do you think?

Sure thing, could you give some advice on the format/template for such a discussion? Or is it just as simple as starting a discussion and saying "we plan to deprecate this"?

samueljseay replied:
Deprecation resolved: https://github.com/woocommerce/woocommerce/pull/45974/commits/f9f4d43aae58e7ce849ef5707a60b2ecdc276907

ralucaStan replied:
> If we continue to support icon I think it implies we'd support the rest of the icon api supported on WP button as well, such as iconPosition and iconSize.

If there are users of this WP icon API via the Button component, what options would they have in the future to achieve equivalent output?

As the surface area is big, and extensions might be using this button in their implementations, I am a bit concerned with deprecating this `icon` prop without sharing ways [code examples] to achieve the same behavior.

Ideally, we should map the current API for the 1st phase and document improvements for future work on the components. 

These deprecations could be announced ahead of time with a communicated removal in a specific WooCommerce version.

samueljseay replied:
> Ideally, we should map the current API for the 1st phase and document improvements for future work on the components.

It's just that this usage was never documented or suggested in Storybook. So what could be considered the "API". props like `icon` and `iconPosition` and such are imo API of WP button but not API of this button. It's murky though I know. I can understand your concern.

> If there are users of this WP icon API via the Button component, what options would they have in the future to achieve equivalent output?

>As the surface area is big, and extensions might be using this button in their implementations, I am a bit concerned with deprecating this icon prop without sharing ways [code examples] to achieve the same behavior.

That part is pretty easy, rendering an icon as a child instead like I have done here in this LOC

samueljseay commented on plugins/woocommerce-blocks/assets/js/base/components/drawer/style.scss:
```diff
@@ -140,9 +145,6 @@ $drawer-animation-duration: 0.3s;
 		outline: none;
 	}
 
-	> span {
```
We disable the text wrapper in this button via `removeTextWrap` which I introduced, so this is not needed now.

samueljseay commented on plugins/woocommerce-blocks/assets/js/base/components/drawer/style.scss:
```diff
@@ -112,7 +117,7 @@ $drawer-animation-duration: 0.3s;
 
 // Important rules are needed to reset button styles.
-.wc-block-components-drawer__close {
+.wc-block-components-button.wc-block-components-drawer__close {
```
To ensure specificity I've appended the button class. In future I think we could refactor the base button styles, but it's a lot more risk and complexity to do that in the scope of this PR and it would have wide-ranging style implications.

samueljseay commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/style.scss:
```diff
@@ -34,6 +35,11 @@
 		outline: 1px solid $universal-border-strong;
 		background-color: $universal-background;
+
+		&:focus {
```
all focus and hover styles are already being overridden so I have not made something **less** accessible with these changes imo, this just ensures they are specific enough in the `selected` to not change the appearance from the existing implementation.

ralucaStan replied:
There are some changes to the styles in this PR https://github.com/woocommerce/woocommerce/issues/46008. Maybe we should wait for that to be merged?

opr replied:
Hi @ralucaStan just to add more info - the PR you're referring to is #46150 and it is already merged.

I rebased this PR locally on trunk and it works fine with the changes from that PR 👌🏼

samueljseay replied:
Thanks for testing that @opr 🙇🏻

nielslange replied:
> Hi @ralucaStan just to add more info - the PR you're referring to is https://github.com/woocommerce/woocommerce/pull/46150 and it is already merged.

@opr The issue #46008, which @raluca referred to, respectively the corresponding PR #46362 actually contains CSS changes that affect the radio buttons, e.g. the outline width on focus will be increased from 1px to 1.5px. 

@samueljseay Based on https://github.com/woocommerce/woocommerce/pull/45974#pullrequestreview-1991233348, I'd say, let's merge this PR first. I'll then rebase my PR and make sure all desired styles are applied to the updated UI elemenets.

opr replied:
Ah my bad for the misunderstanding, sorry! Ok then it _does_ make sense to wait for that to be merged, or rebase locally and see if it works.

Thanks for clarifying this @nielslange 🙇🏼

samueljseay replied:
@nielslange I agree probably to merge this PR first makes most sense imo, just because it's actual structure changes. If you need any help to test the changes in your PR later, please feel free to loop me in.

samueljseay commented on plugins/woocommerce-blocks/bin/webpack-helpers.js:
```diff
@@ -97,6 +97,8 @@ const getAlias = ( options = {} ) => {
 			__dirname,
 			`../assets/js/templates/`
 		),
+		'react/jsx-dev-runtime': require.resolve( 'react/jsx-dev-runtime.js' ),
```
Needed for ariakit to resolve the jsx runtime.

samueljseay committed:
It was not correct to have role button on link with href.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Fix lint issues.

samueljseay committed:
Adjust e2e test now that links are always links.

samueljseay committed:
Fix one more lint css issue.

samueljseay committed:
Remove unused import of wp components radio

samueljseay committed:
Experiment: try allowing for undefined errors thrown in error boundary to deal with Jest test issue.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Revert changes to block error boundary.

samueljseay committed:
Fix typing issues in button

samueljseay committed:
Change assertion in inserter that seems to be incorrect based on block editor behaviour.

opr commented on plugins/woocommerce-blocks/assets/js/base/components/button/index.tsx:
```diff
@@ -23,48 +33,76 @@ export interface ButtonProps
-		variant,
-		{
-			'wc-block-components-button--loading': showSpinner,
+const Button = forwardRef< HTMLButtonElement, ButtonProps | LinkProps >(
```
I think we need to do 
```suggestion
const Button = forwardRef< HTMLButtonElement, ButtonProps & LinkProps >(
```

here because otherwise it results in TS errors for me: `TS2339: Property removeTextWrap does not exist on type PropsWithChildren<LinkProps | ButtonProps>` however that seems to make `href` a required prop. We could then make href optional since it's not always going to be a link.

samueljseay replied:
@opr sorry I think you reviewed at an inopportune time while I was messing with the types. I accidentally removed LinkProps extending from ButtonProps. 

Resolved https://github.com/woocommerce/woocommerce/pull/45974/commits/0ee80b21ca698796dd29c213637b35e488e76f38

opr commented on plugins/woocommerce-blocks/assets/js/base/components/button/index.tsx:
```diff
@@ -23,48 +33,76 @@ export interface ButtonProps
-		<WPButton className={ buttonClassName } { ...props }>
-			{ showSpinner && <Spinner /> }
+		const buttonChildren = removeTextWrap ? (
+			props.children
```
If `href` is not optional (and the change I suggested above is applied) this gives an error `TS2339: Property children does not exist on type never` - I think it's because `'href' in props` always returns true.

samueljseay replied:
See https://github.com/woocommerce/woocommerce/pull/45974#discussion_r1558619128

opr reviewed (changes requested):
Thanks @samueljseay I have tested the changes out and there doesn't seem to be any visual or functional regressions in terms of user experience.

I do notice some failing E2E tests, however maybe this just needs a rebase?

I left a couple of comments about TS too.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Fix type issues.

samueljseay committed:
Add a deprecation warning for the usage of showSpinner.

samueljseay committed:
Revert e2e test change.

samueljseay committed:
Ignore console warnings from recent deprecation.

samueljseay commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/test/block.js:
```diff
@@ -104,6 +104,10 @@ describe( 'Testing cart', () => {
 		render( <CheckoutBlock /> );
+
+		// TODO: Fix a recent deprecation of showSpinner prop of Button called in this component.
+		expect( console ).toHaveWarned();
```
`@wordpress/jest-console` automatically causes all tests with warnings about the new deprecation to fail. I consider this TODO with assertion to be the most pragmatic way to fix this for now. We could follow up in a cooldown with a clean up task to have these components render the spinner instead.

samueljseay committed:
Move the console warning ignore to the correct place.

samueljseay commented:
@opr This is ready for re-review.

The "JavaScript E2E Tests" are flaky and unrelated. I'm seeing inserter tests fail on other branches too such as: https://github.com/woocommerce/woocommerce/actions/runs/8625311160/job/23641705278?pr=46402

This exact failure is different from my last run, but I saw this exact set of failures on a previous commit, so I'm certain these are just flaky tests.

opr reviewed (approved):
Thanks for making the changes, @samueljseay, tests are passing and the code changes since last review look good! Let's 🚢 this 🥳

ralucaStan commented on plugins/woocommerce-blocks/assets/js/base/components/cart-checkout/totals/coupon/test/index.tsx:
```diff
@@ -31,6 +32,9 @@ describe( 'TotalsCoupon', () => {
 		} );
 		rerender( <TotalsCoupon instanceId={ 'coupon' } /> );
+
+		// TODO: Fix a recent deprecation of showSpinner prop of Button called in this component.
```
Could you provide some context for this deprecation?

samueljseay replied:
You mean like add a link to the code where it's deprecated? I added this deprecation as part of this PR.

ralucaStan replied:
I wanted to learn more about the reasons behind the deprecation. Could you mention them in the description PR? It should be announced at that level as well.

samueljseay replied:
@ralucaStan absolutely. I've outlined my reasoning in this comment: https://github.com/woocommerce/woocommerce/pull/45974#discussion_r1556928627

Do you feel this covers my reasoning well enough? If so I'll add it to the description.

ralucaStan replied:
That comment shows that you aren't going to deprecate, but the discussion evolved and the decision was made to deprecate.

I'd be explicit in the description of the PR about the 2 deprecations introduced by this PR and the workarounds needed to accomplish the same behavior.

ralucaStan replied:
I also recommend creating a ticket for actually deprecating the `showSpinner` property. In this case I think it's ok to recommend composability.

For the rest of the components from the project, I recommend documenting the nice to have improvements in new tickets and not deprecating parts of the API, even if they might make sense.

 I think it's important we focus on replacing the components and gathering nice to haves and suggestions for improving the APIs. This will mean little friction for our consumers and for us.

samueljseay replied:
I've updated the description to try explain both removal of icon and deprecation of showSpinner

samueljseay replied:
> I also recommend creating a ticket for actually deprecating the showSpinner property. In this case I think it's ok to recommend composability.

> For the rest of the components from the project, I recommend documenting the nice to have improvements in new tickets and not deprecating parts of the API, even if they might make sense.

I can see your view point, but why are we uncomfortable to deprecate anything? I think we should continue to improve things in place as we work on them personally, and deprecation has only been done where it's very low effort to do so. I think there is a tendency to write tickets for everything and dump them into a cooldown queue and never get to them? I think we should always be leaving the code better than we found it.

ralucaStan replied:
I might have needed to be more explicit. I am not against deprecations. I was only asking to create a ticket for the actual removal of that code in the future. If we don't have a removal ticket then we risk living with that deprecation code hiding in the code base and not being intentional about removing it at some point.

samueljseay replied:
@ralucaStan Thanks I understand you now! 🙏🏻 

Here's an issue to track the removal: https://github.com/woocommerce/woocommerce/issues/46812

> For the rest of the components from the project, I recommend documenting the nice to have improvements in new tickets and not deprecating parts of the API, even if they might make sense.

This still sounds to me like you don't want us to do any actual deprecation for any more parts of this project though? I still think it would be a mistake to approach our code that way, but I'll keep this in mind as we work on the remaining components.

samueljseay replied:
> I'd be explicit in the description of the PR about the 2 deprecations introduced by this PR and the workarounds needed to accomplish the same behavior.

@ralucaStan yes sorry, I did intend to expand on what I had said in the comment. I've now expanded and given detail of things like workarounds to achieve the same result.

samueljseay replied:
I think this is resolved now.

ralucaStan commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx:
```diff
@@ -178,15 +185,16 @@ const Block = ( {
 
 	return (
-		<RadioGroup
+		<div
```
should we keep some css classes on this div that belonged to the RadioGroup?

samueljseay replied:
Are you suggesting that CSS classes are public API? The classes aren't needed for any other reason.

ralucaStan replied:
In the checkout blocks they are public APIs, yes

samueljseay replied:
@ralucaStan I'm just wondering in that case how do we ever deprecate a css class? 

I'm just wondering how strictly we really adhere to this? I'm happy to do whatever is needed here, but I think it's unwise to commit to not changing any CSS classes, especially as HTML will change in structure too, if someone targets a class that is no longer a descendant we wouldn't support that would we? How do we draw a sensible line here?

ralucaStan replied:
In general we kept track of classes using the prefix `.wc-block-components`; we also have some legacy classes `wc-block`. 

RadioGroup is not part of this, but I still think it should be communicated in the release announcement. @opr what do you think? Do you have more context on the CSS classes being public API?

samueljseay replied:
@ralucaStan just updating to say, As much as it pains me, I'll add whatever classes from the old wp component that we have to.

I wish we could deprecate classes somehow, but I think for now it's better to play it safe. Thanks for challenging this one.

samueljseay replied:
There was only one class to add to the div. I've added it with a comment here: https://github.com/woocommerce/woocommerce/pull/45974/commits/dd408abc341084b669632d9edfaff5a0723e40c8

ralucaStan commented on plugins/woocommerce-blocks/assets/js/base/components/button/index.tsx:
```diff
@@ -1,70 +1,127 @@
+			return (
+				<AriakitButton
+					render={
+						<a
```
are the CSS classes automatically added to the anchor?

samueljseay replied:
@ralucaStan yes they are. Ariakit is a bit magic but it merges all the props passed onto the component passed to `render` See https://ariakit.org/guide/composition#merging-the-rendered-element-props

ralucaStan replied:
Nice!

ralucaStan reviewed (commented):
I haven't tested the branch yet, but I'm submitting my review to get the conversation started.

samueljseay committed:
Add old component class to div for back compat.

samueljseay commented:
@opr @ralucaStan @nielslange 

I believe I've addressed everything that was asked for.

* Added more detailed discussion of deprecations with suggested alternatives
* Added classname from radio group to the div where it was previously used
* Created ticket for the removal of deprecated `showSpinner`

If any of y'all would like to re-review I would appreciate it. 🙇🏻

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

ralucaStan commented:
@samueljseay great work on this PR. As I'm testing I noticed there aren't testing instructions for the Radio and Radio group. It would be great to add those as well.

ralucaStan commented:
I am testing using as well keyboard navigation and voice control; 
I have to say that the keyboard navigation is definitely better with Aria kit; For example the Local pickup options can be selected with space when navigating through them; 

I've tested Chrome, Firefox, Safari, Edge and the UI looks ok 👌 


**Mini cart close button (top right of open mini-cart)**

works as expected; I noticed the `components-button` is missing; I'm curious were that was coming from. I don't think we should add it back as it has already blocks specific classes.

 **Cart block -> shipping -> label of radio inputs**

The font size seems bigger; I'd also look at different screen sizes, there is some overlap at one point between the button and the text 

Now:
<img width="393" alt="Screenshot 2024-04-26 at 11 54 36" src="https://github.com/woocommerce/woocommerce/assets/1628454/28c6fb9e-073c-40f5-b78d-801cde15b248">
Before:
<img width="368" alt="Screenshot 2024-04-26 at 11 55 46" src="https://github.com/woocommerce/woocommerce/assets/1628454/e6f62636-f4e8-4303-801c-330d70abeb2b">


**Cart Block -> Coupon form**
I like how the disabled buttons don't react to hover, that's an improvement; however we don't have a disabled visual state; could you create a ticket for adding this? or sync with @nikkivias on defining the disabled state for buttons?


<img width="384" alt="Screenshot 2024-04-26 at 12 02 48" src="https://github.com/woocommerce/woocommerce/assets/1628454/94241aa7-7338-46dc-a98a-072a3042fbe6">

Also, I wonder if aria-disabled="true" and disabled are required at the same time, but it looks like AriaKit is handling all the logic here.
 
```
<button aria-disabled="true" type="submit" class="wc-block-components-button wp-element-button wc-block-components-totals-coupon__button contained" disabled="" style="pointer-events: none;"><span class="wc-block-components-button__text">Apply</span></button>
```

**Checkout block**
This is a pretty serious bug: When checking out as a guest user I was able to place an order without a shipping method selected;
We usually select the first shipping option, but on this branch the option was not selected by default.

> All buttons and link buttons in the cart and checkout
Would be helpful to list the elements that aren't obviously buttons.

> Close button on the notice component

Can you provide some testing steps to see this notice?

ralucaStan reviewed (changes requested):
I added some findings; I'll continue testing in case I missed some elements from the blocks;

samueljseay commented:
@ralucaStan Thanks so much for the detailed testing! It's super helpful.

> As I'm testing I noticed there aren't testing instructions for the Radio and Radio group. It would be great to add those as well.

I added one note, with not enough detail sorry. The only thing to test related to RadioGroup is this feature:

<img width="857" alt="Screenshot 2024-04-29 at 10 38 21 AM" src="https://github.com/woocommerce/woocommerce/assets/1281828/fed31041-60ce-49b6-aecc-036699848721">

Those 2 "radio's" are now ariakit buttons. There is no other usage of the radio/radiogroup

>The font size seems bigger; I'd also look at different screen sizes, there is some overlap at one point between the button and the text

odd since I haven't actually changed these but removing or adjusting some style could have affected them. 

> however we don't have a disabled visual state; could you create a ticket for adding this? or sync with @nikkivias on defining the disabled state for buttons?

Sure, I'll do this.

> Also, I wonder if aria-disabled="true" and disabled are required at the same time, but it looks like AriaKit is handling all the logic here.

yes you absolutely must have both. aria disabled is just a cue for assistive technologies but does not make a button "non-interactable" which is the job of disabled attribute.

> This is a pretty serious bug: When checking out as a guest user I was able to place an order without a shipping method selected; but on this branch the option was not selected by default.
We usually select the first shipping option, see here

Quite strange again since I have not adjusted the radio inputs, but I will try and see what is going on!

> Can you provide some testing steps to see this notice?

Good point, I'll add those. thanks again for the detailed review.

samueljseay committed:
Merge remote-tracking branch 'origin/trunk' into dev/introduce-ariakit-button

samueljseay commented:
<img width="549" alt="Screenshot 2024-04-29 at 11 05 58 AM" src="https://github.com/woocommerce/woocommerce/assets/1281828/00e483f1-bcc9-4bcc-86bb-10b28c7ab7c9">

@ralucaStan I think there is something odd in your environment. I cannot reproduce this, that is to say, the font is always larger. The font appears larger on trunk and my branch. I checked this several times to sanity check. It's also larger for the radios on the checkout page too. Although it could be theme related, but by default on TT4 they use the WP medium font size, which equates to 16.8px.

Also worth mentioning I didn't do anything to radio's (my description about that was misleading. As I mentioned in previous comment, i only adjusted the radiogroup responsible for those 2 options i screenshotted). so it's at least unlikely that anything I did here affected this.

samueljseay commented:
> This is a pretty serious bug: When checking out as a guest user I was able to place an order without a shipping method selected; but on this branch the option was not selected by default.
We usually select the first shipping option, see here

@ralucaStan I also cannot reproduce this. But I could see how its possible my changes affected it. Could you please provide detailed reproduction steps.

samueljseay commented:
Issue to track the addition of visual state: https://github.com/woocommerce/woocommerce/issues/47008

ralucaStan commented:
> Maybe I need more detailed repro steps if there is an edge case I'm missing here.

Sure thing; I have this setup for my store: 
- 1 shipping rate; flat rate for Europe
- 1 Local pickup option defined
- store country US, Default customer location: shop country

- Go on an incognito window; 
- Add product to cart
- Go directly to Checkout block; for me local pickup is selected by default
- switch to shipping
- change country from US to France
- Notice the shipping option appears and it's not selected

Let me know if that helps

ralucaStan commented on plugins/woocommerce-blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-method-block/block.tsx:
```diff
@@ -128,8 +132,10 @@ const ShippingSelector = ( {
 
 	return (
-		<Radio
-			value="shipping"
```
@samueljseay I noticed that for the previous radio you keep the `value="pickup"` whereas here you remove the `			value="shipping"`

samueljseay replied:
@ralucaStan thank you for spotting this. The value isn't needed for either because it's handled in creating the on click handler. i'll remove the other value for consistency

samueljseay replied:
Resolved in https://github.com/woocommerce/woocommerce/pull/45974/commits/225809380d185b093bea7137e1654b13ac793526

samueljseay commented:
@ralucaStan following your steps I can reproduce that behaviour on trunk. I think we should raise a bug.

Maybe there was a caching issue in your environment but I both these behaviours I see happening on trunk. I can also reproduce this issue for a non guest user.

samueljseay committed:
Remove redundant value from button that is not needed.

samueljseay commented:
I've marked this as needing dev note, because we should inform devs that using the Button with icon will require some adjustment to their implementations to render icon as a child.

ralucaStan commented:
> @ralucaStan following your steps I can reproduce that behaviour on trunk. I think we should raise a bug.

Thank you for confirming this and diving deeper into the issue. I agree that some front-end validation should be in place in case the auto-selection fails. 

Let me know if you can open a ticket and surface your findings there as well.

samueljseay commented:
@ralucaStan apart from raising that issue can we merge this?

ralucaStan reviewed (approved):
I created a ticked for the shipping options bug https://github.com/woocommerce/woocommerce/issues/47115
